src/pagination/index.tsx

```diff
@@ -84,105 +114,158 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
 
 	if (pageSizes !== undefined && pageSizes !== icache.get('pageSizes')) {
 		icache.set('pageSizes', pageSizes);
-		icache.set('pageSizesData', pageSizes.map((ps) => ({ value: ps.toString() })));
+		icache.set(
+			'pageSizesResource',
+			createMemoryResourceWithData(pageSizes.map((ps) => ({ value: ps.toString() })))
+		);
+	}
+
+	if (theme.get() !== icache.get('theme')) {
+		icache.set('theme', theme.get());
+		icache.delete('nodeWidths');
 	}
 
 	const page = icache.getOrSet('currentPage', 1);
 	const pageSize = icache.getOrSet('pageSize', 10);
-	const pageSizesData = icache.get('pageSizesData');
-
-	const leading: WNode[] = [];
-	const leadingCount = Math.min(siblingCount, page - 1);
-	for (let i = 1; i <= leadingCount; i++) {
-		const linkTo = page - i;
-
-		leading.unshift(
-			<button
-				type="button"
-				key={`numberedLink-${linkTo}`}
-				classes={[classes.numberedLink, classes.link]}
-				onclick={(e) => {
-					e.stopPropagation();
-					icache.set('currentPage', linkTo);
-					onPage && onPage(linkTo);
-				}}
-			>
-				{linkTo.toString()}
-			</button>
-		);
-	}
+	const pageSizesResource = icache.get('pageSizesResource');
 
-	const trailing: WNode[] = [];
-	const trailingCount = Math.min(siblingCount, total - page);
-	for (let i = 1; i <= trailingCount; i++) {
-		const linkTo = page + i;
-
-		trailing.push(
-			<button
-				type="button"
-				key={`numberedLink-${linkTo}`}
-				classes={[classes.numberedLink, classes.link]}
-				onclick={(e) => {
-					e.stopPropagation();
-					icache.set('currentPage', linkTo);
-					onPage && onPage(linkTo);
-				}}
-			>
-				{linkTo.toString()}
-			</button>
-		);
+	const showPrev = page > 1;
+	const prevLink = (
+		<button
+			type="button"
+			key="prev"
+			classes={[classes.prev, classes.link]}
+			onclick={(e) => {
+				e.stopPropagation();
+				icache.set('currentPage', page - 1);
+				onPage && onPage(page - 1);
+			}}
+		>
+			<div classes={classes.icon}>
+				<Icon type="leftIcon" />
+			</div>
+			<div classes={classes.label}>{messages.previous}</div>
+		</button>
+	);
+
+	const showNext = page < total;
+	const nextLink = (
+		<button
+			type="button"
+			key="next"
+			classes={[classes.next, classes.link]}
+			onclick={(e) => {
+				e.stopPropagation();
+				icache.set('currentPage', page + 1);
+				onPage && onPage(page + 1);
+			}}
+		>
+			<div classes={classes.icon}>
+				<Icon type="rightIcon" />
+			</div>
+			<div classes={classes.label}>{messages.next}</div>
+		</button>
+	);
+
+	const leadingSiblings: RenderResult[] = [];
+	const trailingSiblings: RenderResult[] = [];
+	const containerWidth = dimensions.get('links').size.width;
+	if (containerWidth) {
+		const nodeWidths = icache.getOrSet('nodeWidths', () => ({
```

Comment:
I think this would be better as individual icache entries.
Is it possible to just get the widget width when rendered with zero pages, and the width of a single page link and use those to do the maths? This seems like a fair few things to be checking. 
It also appears that it's duplicating the render code for the buttons / pages etc which could easily get out of sync.

Reply:
The difficulty is that "zero page width" changes for the first and last pages because there are no previous and next links, respectively.

Reply:
I'll clean up the whole loop and we can revisit.